File: packages/core/src/agent.ts
import type { SandboxId } from "./ids.js";
import type { SandboxProvider } from "./sandbox.js";

/**
 * Drives a headless coding agent inside a sandbox. We do NOT reimplement the
 * agent loop — for Claude Code this shells out to `claude -p ... --resume`
 * (see PLAN.md §6). Other CLIs are a different runner, not a rewrite.
 */

export type AgentEvent =
  | { readonly type: "session"; readonly agentSessionId: string }
  | { readonly type: "text"; readonly text: string }
  | { readonly type: "tool_use"; readonly name: string; readonly input: unknown }
  | { readonly type: "tool_result"; readonly name: string; readonly ok: boolean }
  | { readonly type: "error"; readonly message: string }
  | { readonly type: "done"; readonly exitCode: number };

export interface RunRequest {
  readonly sandboxId: SandboxId;
  readonly prompt: string;
  /** Working directory inside the box (the seeded workspace). */
  readonly cwd: string;
  /**
   * Resume an existing agent conversation by its id (Claude's session_id).
   * Omit to start a fresh conversation.
   */
  readonly resumeAgentSessionId?: string;
  /** Allowed tools, passed through to the agent CLI. */
  readonly allowedTools?: readonly string[];
  readonly env?: Readonly<Record<string, string>>;
  readonly signal?: AbortSignal;
}

export interface RunResult {
  /** The agent's own session id, captured for later --resume. */
  readonly agentSessionId: string;
  readonly exitCode: number;
}

export interface AgentRunner {
  /** Human-readable runner name, e.g. "claude-code". */
  readonly name: string;
  /**
   * Run one turn (or a whole task) non-interactively, streaming events.
   * Resolves when the agent process exits.
   */
  run(
    provider: SandboxProvider,
    req: RunRequest,
    onEvent: (e: AgentEvent) => void,
  ): Promise<RunResult>;
}

File: crates/core/src/agent.rs
//! Drives a headless coding agent inside a sandbox. We do NOT reimplement the
//! agent loop; for Claude Code this shells out to `claude -p ... --resume` (see
//! PLAN.md section 6). Other CLIs are a different runner, not a rewrite.

use std::collections::HashMap;

use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use tokio::sync::mpsc;

use crate::errors::Result;
use crate::ids::SandboxId;
use crate::sandbox::SandboxProvider;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AgentEvent {
    /// The agent's own conversation id, captured for later resume.
    Session { agent_session_id: String },
    Text { text: String },
    ToolUse { name: String, input: serde_json::Value },
    ToolResult { name: String, ok: bool },
    Error { message: String },
    Done { exit_code: i64 },
}

#[derive(Debug, Clone)]
pub struct RunRequest {
    pub sandbox_id: SandboxId,
    pub prompt: String,
    /// Working directory inside the box (the seeded workspace).
    pub cwd: String,
    /// Resume an existing agent conversation by id (Claude session_id). None
    /// starts a fresh conversation.
    pub resume_agent_session_id: Option<String>,
    /// Allowed tools, passed through to the agent CLI.
    pub allowed_tools: Vec<String>,
    pub env: HashMap<String, String>,
}

#[derive(Debug, Clone)]
pub struct RunResult {
    /// The agent's own session id, captured for later resume.
    pub agent_session_id: Option<String>,
    pub exit_code: i64,
}

#[async_trait]
pub trait AgentRunner: Send + Sync {
    /// Human readable runner name, e.g. "claude-code".
    fn name(&self) -> &str;

    /// Run one task non interactively, streaming events to `events`. Resolves
    /// when the agent process exits.
    async fn run(
        &self,
        provider: &dyn SandboxProvider,
        req: RunRequest,
        events: mpsc::Sender<AgentEvent>,
    ) -> Result<RunResult>;
}
